CMakeLists.txt

```diff
@@ -510,7 +507,7 @@ target_include_directories(precice PUBLIC
   $<INSTALL_INTERFACE:include>
   )
 
-target_link_libraries(precice PRIVATE preciceCore)
+target_link_libraries(precice PRIVATE preciceCore preciceExternal)
```

Comment:
I still think it would be way simpler to extend the preciceCore directly instead of introducing multiple new targets. This way there is no risk for linking issues and the overall CMake code shrinks. Furthermore, symbol hiding (which doesn't exist in static libraries) and wider LTO scope should lead to more optimization potential.

Reply:
How would this then look like? Note that this would then lead to the whole preciceCore linking against Kokkos and Kokkos ships its own compiler launcher, such that preCICE inherits flags from Kokkos. Different source code files still need to be compiled by different compilers. In principle, we can of course also build the helper interfaces as dynamic libraries, but then we might need to install them.

Reply:
> Note that this would then lead to the whole preciceCore linking against Kokkos and Kokkos ships its own compiler launcher, such that preCICE inherits flags from Kokkos. Different source code files still need to be compiled by different compilers.

If kokkos defines the compiler launcher for CUDA, then this shouldn't impact precice sources.

> How would this then look like?

I see two options:

1. we use `target_sources` to add all extra files straight to the preciceCore, link the core to ginkgo and Kokkos, and let CMake figure out the rest. If this works, then this would be the simplest solution.
2. we define an object library for device code, link that to ginkgo and Kokkos, and link that to the preciceCore. precice and testprecice may need to be manually linked to ginkgo and Kokkos for this to work.

https://cmake.org/cmake/help/latest/command/add_library.html#object-libraries

Reply:
> If kokkos defines the compiler launcher for CUDA, then this shouldn't impact precice sources.

The launcher is used for all files. Kokkos can also be used for other optimizations (OpenMP, Serial, Vectorization etc..)

Option two is what we do right now, correct?

How would option one look like with the additional variables and language definitions?

Reply:
Let me try this

Reply:
Tried something like this

```diff
 # Setup Ginkgo
 if (PRECICE_FEATURE_GINKGO_MAPPING)
   # Next we configure the common kernels for the matrix assembly
-  add_library(preciceCore_ginkgo STATIC
-          src/mapping/device/GinkgoRBFKernels.cpp
-          src/mapping/device/Ginkgo.cpp)
-  target_link_libraries(preciceCore_ginkgo PRIVATE Ginkgo::ginkgo Kokkos::kokkos preciceCore)
+
+  target_sources(preciceCore PRIVATE
+  src/mapping/device/GinkgoRBFKernels.cpp
+  src/mapping/device/Ginkgo.cpp
+  )
+
+  target_link_libraries(preciceCore PRIVATE Ginkgo::ginkgo Kokkos::kokkos)
   target_compile_definitions(preciceCore PUBLIC EIGEN_NO_CUDA)
-  target_compile_definitions(preciceCore_ginkgo PUBLIC EIGEN_NO_CUDA)
-  set_target_properties(preciceCore_ginkgo PROPERTIES POSITION_INDEPENDENT_CODE ON)
 
   # Disable compiler launcher since Kokkos ships its own nvcc wrapper
-  set_target_properties(preciceCore_ginkgo PROPERTIES CXX_COMPILER_LAUNCHER "")
+  set_target_properties(preciceCore PROPERTIES CXX_COMPILER_LAUNCHER "")
 
```

but this triggers all kinds of errors in the 'preCICE' TUs, especially Boost PP macros seem to be a problem here:

```bash
precice/src/mesh/Vertex.hpp(130): error: identifier "PRECICE_ASSERT_IMPL_BOOST_PP_IIF_BOOST_PP_BOOL_BOOST_PP_IIF_BOOST_PP_NOT_EQUAL_CHECK_BOOST_PP_NOT_EQUAL_BOOST_PP_VARIADIC_SIZE" is undefined

precice/src/mesh/Vertex.hpp(130): error: identifier "BOOST_PP_NIL" is undefined
```

Reply:
https://github.com/precice/precice/pull/2051#discussion_r1682815113 this looks correct and the errors look very strange.

Then we need option two.

> Option two is what we do right now, correct?

Currently, we build static libraries which are then linked against the shared library. This comes with some additional implications.

Using an object library means "please compile these sources in this different way, I'll use them later to link a library.". This is exactly what we need here.

Reply:
Can you give me some hint on what to do in this case? Changing `STATIC` to `OBJECT`, yes, but I also need to link Kokkos and other TPLs then against the interface, right?

```diff
-  add_library(preciceCore_ginkgo STATIC
+  add_library(preciceCore_ginkgo OBJECT
```
and
```diff
-  target_link_libraries(preciceExternal INTERFACE preciceCore_ginkgo)
-else()
+  target_link_libraries(preciceExternal INTERFACE $<TARGET_OBJECTS:preciceCore_ginkgo>)
+  target_link_libraries(preciceExternal INTERFACE Kokkos::kokkos)
```

Reply:
After changing to `OBJECT`, you should be able to directly link to the object lib. Older version of CMake didn't do this correctly, but 3.12 should work.

```diff
-  target_link_libraries(preciceExternal INTERFACE preciceCore_ginkgo)
+  target_link_libraries(preciceCore PUBLIC preciceCore_ginkgo)
+  target_link_libraries(preciceCore_ginkgo PUBLIC Kokkos::kokkos)
```

Reply:
If I do this I end up with the following:

```bash
CMake Error: The inter-target dependency graph contains the following strongly connected component (cycle):
  "preciceCore" of type OBJECT_LIBRARY
    depends on "preciceCore_ginkgo" (weak)
  "preciceCore_ginkgo" of type OBJECT_LIBRARY
    depends on "preciceCore" (weak)
At least one of these targets is not a STATIC_LIBRARY.  Cyclic dependencies are allowed only among static libraries.
```
so I should link the ginkgo Object against preCICE itself?

Reply:
What could work is to remove the dependency of Core on Core_ginkgo and make precice and testprecice depend on both Core and Core_ginkgo.

The best solution would be https://github.com/precice/precice/pull/2051#discussion_r1682815113 though. Maybe its worth having a deeper look into why this doesn't work. The problem is the Kokkos compiler wrapper, correct?

Reply:
I made it to work (see the last two commits) and compiling succeeds for all backends we offer. The culprit of the aforementioned issues were actually, that nvcc has issues with variadic macros.  I enabled them now in all TUs in preciceCore `BOOST_PP_VARIADICS`. Also, for HIP and Cuda, I had to resolve the `::min` conflict in CompositionalCouplingSchemes. 

For testing, the OpenMP setup works flawless. However, for Cuda and Hip, I get the following runtime error:

```bash
 Product of tile dimensions exceed maximum limit: 0
ExecSpace Error: MDRange tile dims exceed maximum number of threads per block - choose smaller tile dims
Backtrace:
[0x7fc513d50ad5] Kokkos::Impl::save_stacktrace()
[0x7fc513d24a9c] Kokkos::Impl::host_abort(char const*)
```
which puzzles me a bit. The previous CMake setup works without this error. Any clues here?

Reply:
With Kokkos version 4.3.01, Cuda works and only with Kokkos develop ([eb11070](https://github.com/kokkos/kokkos/commit/eb11070f67565b2e660659f5207f0363bdf3b882)) also HIP works. I have no clue what's going on. With Cuda, version 4.3.00 doesn't work either.

Reply:
Good investigative development